### PR TITLE
HHH-12469 - Add support for IN clause parameter padding to better reuse cached statements

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-12469

### CODE REVIEW

**Comment by vladmihalcea** [comment]
@sebersole @dreab8 When you have time, can you review this PR? Thanks.

File: hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java [hunk 1]
```diff
@@ -1883,4 +1883,22 @@
 	 * @see org.hibernate.query.ImmutableEntityUpdateQueryHandlingMode
 	 */
 	String IMMUTABLE_ENTITY_UPDATE_QUERY_HANDLING_MODE = "hibernate.query.immutable_entity_update_query_handling_mode";
+
+	/**
+	 * By default, the IN clause expends to include all bind parameter values.
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java]: Was curious about the feature so I read the patch: s/expends/expands/

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java]: Thanks for the tip. The IN clause expends quite some resources too :D

File: hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java [hunk 2]
```diff
@@ -528,6 +527,25 @@ public String expandListValuedParameters(String queryString, SharedSessionContra
...
 			final Collection bindValues = entry.getValue().getBindValues();
 
+			int bindValueCount = bindValues.size();
+			int bindValueMaxCount = bindValueCount;
+
+			boolean inClauseParameterPaddingEnabled =
+					session.getFactory().getSessionFactoryOptions().inClauseParameterPaddingEnabled() &&
+					bindValueCount > 2;
+
+			if ( inClauseParameterPaddingEnabled ) {
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java]: Not sure which is ultimately faster, but have you tried `final int bindValueMaxCount = (int) (1L << -Long.numberOfLeadingZeros(count - 1) );`?

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java]: BTW, I got that from Guava Math classes

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java]: Also, no matter which way we choose, I'd much prefer this be moved into a method.  The intent is not very clear in either approach - but calling a method like `#ceilingPowerOfTwo` (Guava's name) is much more readable, imo

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/query/internal/QueryParameterBindingsImpl.java]: I'll move it to some Utils class and check which one is faster as well.

**Pull request merged**